x/validator-preference/msg_server_test.go

```diff
@@ -0,0 +1,124 @@
+package keeper_test
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	valPref "github.com/osmosis-labs/osmosis/v12/x/validator-preference"
+	"github.com/osmosis-labs/osmosis/v12/x/validator-preference/types"
+)
+
+func (suite *KeeperTestSuite) TestSetValidatorSetPreference() {
+	suite.SetupTest()
+
+	// setup 3 validators
+	valAddrs := suite.SetupMultipleValidators(3)
+
+	type param struct {
+		delegator   sdk.AccAddress
+		preferences []types.ValidatorPreference
+	}
+
+	tests := []struct {
+		name       string
+		param      param
+		expectPass bool
+	}{
+		{
+			name: "creation of new validator set",
+			param: param{
+				delegator: sdk.AccAddress([]byte("addr1---------------")),
+				preferences: []types.ValidatorPreference{
+					{
+						ValOperAddress: valAddrs[0],
+						Weight:         sdk.NewDecWithPrec(5, 1),
+					},
+					{
+						ValOperAddress: valAddrs[1],
+						Weight:         sdk.NewDecWithPrec(3, 1),
+					},
+					{
+						ValOperAddress: valAddrs[2],
+						Weight:         sdk.NewDecWithPrec(2, 1),
+					},
+				},
+			},
+			expectPass: true,
+		},
+		{
+			name: "Update existing validator set",
+			param: param{
+				delegator: sdk.AccAddress([]byte("addr1---------------")),
+				preferences: []types.ValidatorPreference{
+					{
+						ValOperAddress: valAddrs[0],
+						Weight:         sdk.NewDecWithPrec(2, 1),
+					},
+					{
+						ValOperAddress: valAddrs[1],
+						Weight:         sdk.NewDecWithPrec(2, 1),
+					},
+					{
+						ValOperAddress: valAddrs[2],
+						Weight:         sdk.NewDecWithPrec(6, 1),
+					},
+				},
+			},
+			expectPass: true,
+		},
+		{
+			name: "create validator set with unknown validator address",
+			param: param{
+				delegator: sdk.AccAddress([]byte("addr2---------------")),
+				preferences: []types.ValidatorPreference{
+					{
+						ValOperAddress: "addr1---------------",
+						Weight:         sdk.NewDec(1),
+					},
+					{
+						ValOperAddress: valAddrs[1],
+						Weight:         sdk.NewDecWithPrec(3, 1),
+					},
+				},
+			},
+			expectPass: false,
+		},
+		{
+			name: "create validator set with weights != 1",
+			param: param{
+				delegator: sdk.AccAddress([]byte("addr3---------------")),
+				preferences: []types.ValidatorPreference{
+					{
+						ValOperAddress: valAddrs[0],
+						Weight:         sdk.NewDecWithPrec(5, 1),
+					},
+					{
+						ValOperAddress: valAddrs[1],
+						Weight:         sdk.NewDecWithPrec(3, 1),
+					},
+					{
+						ValOperAddress: valAddrs[2],
+						Weight:         sdk.NewDecWithPrec(3, 1),
+					},
+				},
+			},
+			expectPass: false,
```

Comment:
The only reason this works right now is because addr3 did not have a previously existing validator set pref. If you choose for example addr1 this would pass (in a bad way) due to the issue explained earlier.

Reply:
Will fix this! :))